tools/offwaketime.py

```diff
@@ -114,58 +114,73 @@ def signal_ignore(signal, frame):
     int t_k_stack_id;
     int t_u_stack_id;
     u32 t_pid;
+    u32 t_tgid;
     u32 w_pid;
-    u32 tgid;
+    u32 w_tgid;
 };
 BPF_HASH(counts, struct key_t);
+
+// Key of this hash is PID of waiting Process,
+// value is timestamp when it went into waiting
 BPF_HASH(start, u32);
+
 struct wokeby_t {
     char name[TASK_COMM_LEN];
     int k_stack_id;
     int u_stack_id;
     int w_pid;
+    int w_tgid;
 };
+// Key of the hash is PID of the Process to be waken, value is information
+// of the Process who wakes it
 BPF_HASH(wokeby, u32, struct wokeby_t);
 
 BPF_STACK_TRACE(stack_traces, STACK_STORAGE_SIZE);
 
 int waker(struct pt_regs *ctx, struct task_struct *p) {
+    // PID and TGID of the target Process to be waken
     u32 pid = p->pid;
     u32 tgid = p->tgid;
 
     if (!(THREAD_FILTER)) {
         return 0;
     }
 
+    // Construct information about current (the waker) Process
     struct wokeby_t woke = {};
     bpf_get_current_comm(&woke.name, sizeof(woke.name));
-
     woke.k_stack_id = KERNEL_STACK_GET;
     woke.u_stack_id = USER_STACK_GET;
-    woke.w_pid = pid;
+    woke.w_pid = bpf_get_current_pid_tgid();
+    woke.w_tgid = bpf_get_current_pid_tgid() >> 32;
+
     wokeby.update(&pid, &woke);
     return 0;
 }
 
 int oncpu(struct pt_regs *ctx, struct task_struct *p) {
-    u32 pid = p->pid, t_pid=pid;
+    // PID and TGID of the previous Process (Process going into waiting)
+    u32 pid = p->pid;
     u32 tgid = p->tgid;
```

Comment:
looks like we do not need tgid assignment here.

Reply:
We don’t use it directly here, but might be used in THREAD_FILTER